bench/bench.py

```diff
@@ -0,0 +1,189 @@
+import statistics
+import time
+
+
+class Bench:
+    name = None
+    driver = None
+    queries = []
+
+    def __call__(self):
+        try:
+            self.setup()
+        except ImportError:
+            print(f'Driver {self.driver} not installed, skipping…')
+            return
+
+        for query in self.queries:
+            self.time(query)
+
+        self.close()
+
+    def time(self, query, iterations=10):
+        times = [None] * iterations
+
+        print(query)
+        for i in range(iterations):
+            start = time.perf_counter()
+            self.run(query)
+            times[i] = time.perf_counter() - start
+
+        median = statistics.median(times)
+        mean = statistics.mean(times)
+        max_ = max(times)
+        min_ = min(times)
+        print(
+            f"Mean: {mean:.4f} - "
+            f"Median: {median:.4f} - "
+            f"Max: {max_:.4f} - "
+            f"Min: {min_:.4f}"
+        )
+        print('---')
+
+    def setup(self):
+        raise NotImplementedError()
+
+    def run(self):
+        raise NotImplementedError()
+
+    def close(self):
+        raise NotImplementedError()
+
+
+class SlonikMixin:
+    driver = 'slonik'
+    conn = None
+
+    def setup(self):
+        import slonik
+
+        self.conn = slonik.Connection.from_env()
+
+        # warmup the connection
+        self.conn.get_one('SELECT 1')
+
+    def close(self):
+        self.conn.close()
+
+
+class PsycopgMixin:
+    driver = 'psycopg2'
+    conn = None
+
+    def setup(self):
+        import psycopg2
+
+        self.conn = psycopg2.connect(host=None)
+        self.cur = self.conn.cursor()
+
+        # warmup the connection
+        self.cur.execute('SELECT 1')
+        self.cur.fetchone()
+
+    def close(self):
+        self.cur.close()
+        self.conn.close()
+
+
+class AsyncpgMixin:
+    driver = 'asyncpg'
+    conn = None
+    loop = None
+
+    def setup(self):
+        import asyncio
+        import asyncpg
+
+        self.loop = asyncio.get_event_loop()
+        self.conn = self.loop.run_until_complete(asyncpg.connect())
+
+        # warmup the connection
+        self.loop.run_until_complete(self.conn.fetch('SELECT 1'))
+
+    def close(self):
+        self.loop.run_until_complete(self.conn.close())
+
+
+# --- One row, several columns ---
+
+
+class ColumnsBench(Bench):
+    name = 'columns'
+    queries = [
+        "SELECT 1",
+        "SELECT 1, 'foo', 42.21::float",
+        "SELECT '3d9d291d-8668-480f-98bf-46ee10d07a5d'::uuid",
+    ]
+
+
+class SlonikColumnsBench(SlonikMixin, ColumnsBench):
+
+    def run(self, query):
+        self.conn.get_one(query)
+
+
+class PsycopgColumnsBench(PsycopgMixin, ColumnsBench):
+
+    def run(self, query):
+        self.cur.execute(query)
+
+
+class AsyncpgColumnsBench(AsyncpgMixin, ColumnsBench):
+
+    def run(self, query):
+        self.loop.run_until_complete(self.conn.fetch(query))
```

Comment:
Do you think the fact you're running `run_until_complete` everytime may have an impact ?

I guess it's about comparing the libs in equal conditions (sync calls) vs in their central usecase.

I'd guess the 2 kinds of benchmarks make sense. A simple python script that has to run n times the queries recieved in stdin, and write the output in stdout, written in each lib's idiomatic way and you measure it with your shell's `time` would make sense too.

Reply:
I wanted this to be as idiomatic as possible for now. I don't think `run_until_complete` really has an overhead compared to a simple `await`, but I'll check.